pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.pinot.core.operator.blocks.IntermediateResultsBlock;
+import org.apache.pinot.core.operator.blocks.TransformBlock;
+import org.apache.pinot.core.operator.transform.TransformOperator;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.DistinctAggregationFunction;
+import org.apache.pinot.core.query.distinct.DistinctTable;
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.index.reader.Dictionary;
+import org.apache.pinot.spi.data.FieldSpec;
+
+import static org.apache.pinot.core.query.distinct.DistinctExecutor.MAX_INITIAL_CAPACITY;
+
+
+/**
+ * Operator which executes DISTINCT operation based on dictionary
+ */
+public class DictionaryBasedDistinctOperator extends DistinctOperator {
+    private static final String OPERATOR_NAME = "DictionaryBasedDistinctOperator";
+
+    private final DistinctAggregationFunction _distinctAggregationFunction;
+    private final Dictionary _dictionary;
+    private final int _numTotalDocs;
+    private final TransformOperator _transformOperator;
+    private IntSet _dictIdSet;
+
+    private IntPriorityQueue _priorityQueue;
+
+    private boolean _hasOrderBy;
+
+    public DictionaryBasedDistinctOperator(IndexSegment indexSegment, DistinctAggregationFunction distinctAggregationFunction,
+                                           Dictionary dictionary, int numTotalDocs,
+                                           TransformOperator transformOperator) {
+        super(indexSegment, distinctAggregationFunction, transformOperator);
+
+        _distinctAggregationFunction = distinctAggregationFunction;
+        _dictionary = dictionary;
+        _numTotalDocs = numTotalDocs;
+        _transformOperator = transformOperator;
+        _dictIdSet = new IntOpenHashSet();
+
+        List<OrderByExpressionContext> orderByExpressionContexts = _distinctAggregationFunction.getOrderByExpressions();
+
+        if (orderByExpressionContexts != null) {
+            OrderByExpressionContext orderByExpressionContext = orderByExpressionContexts.get(0);
+            int limit = _distinctAggregationFunction.getLimit();
+            int comparisonFactor = orderByExpressionContext.isAsc() ? -1 : 1;
+
+            _priorityQueue =
+                    new IntHeapPriorityQueue(Math.min(limit, MAX_INITIAL_CAPACITY), (i1, i2) -> (i1 - i2) * comparisonFactor);
+            _hasOrderBy = true;
+        }
+    }
+
+    @Override
+    protected IntermediateResultsBlock getNextBlock() {
+        int limit = _distinctAggregationFunction.getLimit();
+        ExpressionContext expression = _distinctAggregationFunction.getInputExpressions().get(0);
+
+        assert _distinctAggregationFunction.getType() == AggregationFunctionType.DISTINCT;
+
+        TransformBlock transformBlock = _transformOperator.nextBlock();
```

Comment:
We don't need to read dictIds from the transform block.

- If there is no order-by, read the first `min(limit, dictionarySize)` values from the dictionary
- If there is order-by
  - If dictionary is sorted, read the first if asc, last if desc `min(limit, dictionarySize)` values from the dictionary
  - If dictionary is not sorted, dump all values in the dictionary into the `DistinctTable`

Reply:
Yes, I was not aware that dictionary (when sorted) has IDs strictly in the ascending order of numbers (i.e. starting dictID will always be 0, next 1 ...). Thanks for highlighting that. Fixed per your suggestions.

For the no order by case, we dont really have an order defined in which we read values from the underlying storage (so the order of values read from segment vs dictionary can be different). Hence, need to change DistinctQueriesTest to reflect the same.